### DISCUSSION TITLE
Node.Js - Backend - Building a Compound Response Node.js Server with Routing

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#163] [Tasks]
Task: Building a Compound Response Node.js Server with Routing

Objective: Build a Node.js server that handles HTTP requests, implements routing using Express.js, and generates complex text and JSON responses.

Requirements:
1. Set up a Node.js project and install the required dependencies, including Express.js.
2. Create a new file named `server.js` and set up the basic server configuration.
3. Define the following routes to handle different HTTP request methods and endpoints:

   a) `GET /`: Display a welcome message and provide instructions for accessing different endpoints.
   b) `GET /generate-response`: Generate a complex response containing a combination of text and JSON data.
   c) `POST /process-data`: Receive JSON data in the request body, process it, and generate a response.

4. Implement complex logic within each route to generate dynamic responses. Consider the following requirements for each route:

   a) `/`: Display a welcome message and provide instructions for accessing the other endpoints.
   b) `/generate-response`:
      - Generate a response that includes dynamic information such as the current date and time.
      - Perform complex operations or calculations to generate the required data for the response.
      - Include an array of objects with each object containing various properties.
      - Ensure the response is well-formatted and easy to understand.
   c) `/process-data`:
      - Receive JSON data in the request body.
      - Implement logic to process the data, perform validation, and generate a response.
      - Return an appropriate response based on the processed data.

5. Add error handling middleware to handle any runtime errors or invalid requests.

6. Start the server and test the functionality using a web browser or an API testing tool like Postman.

Additional Tasks:
1. Implement error handling middleware to handle 404 Not Found errors for undefined routes.
2. Add validation logic to the `/process-data` route to ensure the received JSON data is valid.
3. Implement an authentication middleware to protect certain routes and ensure only authorized users can access them.
4. Create a custom middleware that logs each incoming request and its details.
5. Implement additional routes with different HTTP methods and endpoints to showcase more functionality.

Note: This task is designed to be complex and compound, requiring you to demonstrate your understanding of Node.js, Express.js, routing, handling different request methods, generating dynamic responses, and implementing middleware. Feel free to be creative with the logic, data structures, and additional features to showcase your skills and knowledge.

**Comment by sam2elect** [comment]
# CODE
```js

// npm init -y
// npm install http
// Create a new file named index.js in the project directory.
// npm i --g nodemon
// nodemon index.js

const http = require('http');
const url = require('url');

const server = http.createServer((req, res) => {
  const { pathname } = url.parse(req.url);

  if (req.method === 'GET' && pathname === '/') {
    // Display a welcome message and provide instructions for accessing different endpoints
    res.writeHead(200, { 'Content-Type': 'text/plain' });
    res.end('Welcome to the Node.js Server!\n\nAvailable endpoints:\n\n- GET /generate-response\n- POST /process-data');
  } 
  
  else if (req.method === 'GET' && pathname === '/generate-response') {
    // Generate a complex response containing a combination of text and JSON data
    const response = {
      message: 'Welcome to the Complex Response Server!',
      timestamp: new Date().toISOString(),
      data: [
        { name: 'John Doe', age: 30, location: 'New York' },
        { name: 'Jane Smith', age: 28, location: 'San Francisco' }
      ]
    };

    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(response));
  } else if (req.method === 'POST' && pathname === '/process-data') {
    req.on('end', () => {
        const data = JSON.parse(body);
        // Process the received data and generate a response
        const processedData = {
          originalData: data,
          processed: true
        };
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify(processedData));
    } 
)}; else {
    // Handle 404 Not Found errors for undefined routes
    res.writeHead(404, { 'Content-Type': 'text/plain' });
    res.end('404 Not Found');
  }
});

server.listen(5555, () => {
  console.log('Server started on port 5555');
});

```

# API Test

![image](https://github.com/akash-coded/mern/assets/85131631/808e5564-2fe9-41c3-9ddb-02391ae6ac78)
![image](https://github.com/akash-coded/mern/assets/85131631/ca1d31db-88ea-4735-b905-2f04bcb2d167)
![image](https://github.com/akash-coded/mern/assets/85131631/e84bb14e-6897-425e-9bd9-e24b2620c533)
![image](https://github.com/akash-coded/mern/assets/85131631/6f4c20c3-b1ea-4082-9260-4c4c92b90cb8)
![image](https://github.com/akash-coded/mern/assets/85131631/48259b5c-fb48-40e6-8f4e-0ff29dc639d9)

**Comment by pujita789** [comment]
# CODE
```js


const http = require('http');
const url = require('url');

const server = http.createServer((req, res) => {
  const { pathname } = url.parse(req.url);

  if (req.method === 'GET' && pathname === '/') {
    // Display a welcome message and provide instructions for accessing different endpoints
    res.writeHead(200, { 'Content-Type': 'text/plain' });
    res.end('Welcome to the Node.js Server!\n\nAvailable endpoints:\n\n- GET /generate-response\n- POST /process-data');
  }

  else if (req.method === 'GET' && pathname === '/generate-response') {
    // Generate a complex response containing a combination of text and JSON data
    const response = {
      message: 'Welcome to the Complex Response Server!',
      timestamp: new Date().toISOString(),
      data: [
        { name: 'John Doe', age: 30, location: 'New York' },
        { name: 'Jane Smith', age: 28, location: 'San Francisco' }
      ]
    };

    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(response));
  } else if (req.method === 'POST' && pathname === '/process-data') {
    req.on('end', () => {
      const data = JSON.parse(body);
      // Process the received data and generate a response
      const processedData = {
        originalData: data,
        processed: true
      };
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(processedData));
    }
    )
  } else {
    // Handle 404 Not Found errors for undefined routes
    res.writeHead(404, { 'Content-Type': 'text/plain' });
    res.end('404 Not Found');
  }
});

server.listen(2200, () => {
  console.log('Server started on port 2200');
});

**Comment by saugatacr7** [comment]
// npm init -y
// npm install http
// Create a new file named index.js in the project directory.
// npm i --g nodemon
// nodemon server.js

const http = require('http');
const url = require('url');

const server = http.createServer((req, res) => {
  const { pathname } = url.parse(req.url);

  if (req.method === 'GET' && pathname === '/') {
    // Display a welcome message and provide instructions for accessing different endpoints
    res.writeHead(200, { 'Content-Type': 'text/plain' });
    res.end('Welcome to the Node.js Server!\n\nAvailable endpoints:\n\n- GET /generate-response\n- POST /process-data');
  } 
  
  else if (req.method === 'GET' && pathname === '/generate-response') {
    // Generate a complex response containing a combination of text and JSON data
    const response = {
      message: 'Welcome to the Complex Response Server!',
      timestamp: new Date().toISOString(),
      data: [
        { name: 'John Doe', age: 30, location: 'New York' },
        { name: 'Jane Smith', age: 28, location: 'San Francisco' }
      ]
    };

    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(response));
  } else if (req.method === 'POST' && pathname === '/process-data') {
    req.on('end', () => {
        const data = JSON.parse(body);
        // Process the received data and generate a response
        const processedData = {
          originalData: data,
          processed: true
        };
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify(processedData));
    } 
)}; else {
    // Handle 404 Not Found errors for undefined routes
    res.writeHead(404, { 'Content-Type': 'text/plain' });
    res.end('404 Not Found');
  }
});

server.listen(5555, () => {
  console.log('Server started on port 5555');
});

**Comment by starvikas1988** [comment]
```js


const http = require('http');
const url = require('url');

const server = http.createServer((req, res) => {
  const { pathname } = url.parse(req.url);

  if (req.method === 'GET' && pathname === '/') {
    // Display a welcome message and provide instructions for accessing different endpoints
    res.writeHead(200, { 'Content-Type': 'text/plain' });
    res.end('Welcome to the Node.js Server!\n\nAvailable endpoints:\n\n- GET /generate-response\n- POST /process-data');
  }

  else if (req.method === 'GET' && pathname === '/generate-response') {
    // Generate a complex response containing a combination of text and JSON data
    const response = {
      message: 'Welcome to the Complex Response Server!',
      timestamp: new Date().toISOString(),
      data: [
        { name: 'John Doe', age: 30, location: 'New York' },
        { name: 'Jane Smith', age: 28, location: 'San Francisco' }
      ]
    };

    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(response));
  } else if (req.method === 'POST' && pathname === '/process-data') {
    req.on('end', () => {
      const data = JSON.parse(body);
      // Process the received data and generate a response
      const processedData = {
        originalData: data,
        processed: true
      };
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(processedData));
    }
    )
  } else {
    // Handle 404 Not Found errors for undefined routes
    res.writeHead(404, { 'Content-Type': 'text/plain' });
    res.end('404 Not Found');
  }
});

server.listen(2200, () => {
  console.log('Server started on port 2200');
});


```

**Comment by bipin-coded** [comment]
```node
const http = require('http');

const server = http.createServer((req, res) => {

    if (req.url === '/' && req.method === 'GET') {

        res.setHeader(200, 'Content-Type', 'text/html'); 
        res.end('<h3>Welcome to Node JS!</h3>');

    } else if (req.url === '/generate-response' && req.method === 'GET') {

        const data = [
            {
                "id": "0001",
                "type": "donut",
                "name": "Cake",
                "ppu": 0.55,
                "batters": [
                    { "id": "1001", "type": "Regular" },
                    { "id": "1002", "type": "Chocolate" },
                    { "id": "1003", "type": "Blueberry" },
                    { "id": "1004", "type": "Devil's Food" }
                ],
                "topping": [
                    { "id": "5001", "type": "None" },
                    { "id": "5002", "type": "Glazed" },
                    { "id": "5005", "type": "Sugar" },
                    { "id": "5007", "type": "Powdered Sugar" },
                    { "id": "5006", "type": "Chocolate with Sprinkles" },
                    { "id": "5003", "type": "Chocolate" },
                    { "id": "5004", "type": "Maple" }
                ]
            },
            {
                "id": "0002",
                "type": "donut",
                "name": "Raised",
                "ppu": 0.55,
                "batters": [
                    { "id": "1001", "type": "Regular" }
                ],
                "topping": [
                    { "id": "5001", "type": "None" },
                    { "id": "5002", "type": "Glazed" },
                    { "id": "5005", "type": "Sugar" },
                    { "id": "5003", "type": "Chocolate" },
                    { "id": "5004", "type": "Maple" }
                ]
            }
        ];

        const response = {
            message: 'Welcome to the Complex Response Server!',
            timestamp: new Date().toISOString(),
            data: data
        };

        res.setHeader(200, 'Content-Type', 'application/json');
        res.end(JSON.stringify(response, null, 5));

    } else if (req.url === '/process-data' && req.method === 'POST') {

        const params = req.body.params;

        const response = {
            name: params.name,
            email: params.email,
            age: params.age,
            location: params.location
        }

        res.setHeader(200, 'Content-Type', 'application/json');
        res.end(JSON.stringify(response, null, 5));

    } else {

        res.setHeader(404, 'Content-Type', 'application/json');
        res.end('404: Page Not Found');
        
    }

});

server.listen(8000, () => {
    console.log('Listening on port 8000!');
});
```

**Comment by basu-cse** [comment]
const express = require('express')
const app = express()
const port = 3000
const http = require('http');

app.get('/', (req, res) => {
  res.send('Welcome to Node!');
})
app.get('/generate-response', (req, res) => {
  const currentDate = new Date().toISOString();

  const data = [
    {
      name: 'John Doe',
      age: 30,
      location: 'New York',
    },
    {
      name: 'Jane Smith',
      age: 28,
      location: 'San Francisco',
    },
  ];

  const response = {
    message: 'Welcome to the Complex Response Server!',
    timestamp: currentDate,
    data: data,
  };

  res.json(response);
});

app.post('/process-data', (req, res) => {
  // Process JSON data received in the request body
  const data = req.body;

  res.send('Data processed successfully!'); // Send a response message
});

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`)
});

**Reply by raihanul-islam-rakin** [reply]
// Process JSON data received in the request body
const data = req.body;

res.send('Data processed successfully!'); // Send a response message
});

app.listen(port, () => {
console.log(Example app listening on port ${port})
});

**Comment by RevathiSundaramoorthy** [comment]
## Code
```javascript
//load http and url modules
const http = require("http");
const url = require('url');
//define hostname and port
const hostname = "127.0.0.1";
const port = 3002;
//define data variables
const timestamp = new Date();
const data = [
    {
        name: "John Doe",
        age: 30,
        location: "NewYork"
    },
    {
        name: "Jane Smith",
        age: 28,
        location: "San Francisco"
    }
];
const response = {
    message: "Welcome to the Complex Response Server!",
    timestamp: timestamp,
    data: data
};
//define http server with routes
const server = http.createServer((req,res)=>{
//for home page
    if(req.url == "/"){
        res.write("<h1>Welcome to home page<h1>\n");
        res.write("<h2>Go to /generate-response page for dynamic data output</h2>\n");
        res.write("Go to /process-date for processing input");
        res.end();
//to display data
    } else if(req.url == "/generate-response"){
        res.write(JSON.stringify(response));
        res.end();
//to process data from url        
    } else if(req.url.includes("/process-data")){
//parse url to get input objects        
        const q = url.parse(req.url, true);
        const newData = q.query;
//name is a required input        
        if(!newData.name) {
            res.writeHead(500,{"Content-Type":"Text/Plain"});
            res.end("name is required");
        }
//check data array to avoid duplicate entry        
        else {
            const existingname = data.find(person => {return person.name == newData.name});
            if (existingname) {
                res.writeHead(500,{"Content-Type":"Text/plain"});
                res.end("Data already exists")
            }    
            else {
                data.push(newData);
                res.writeHead(200,{"Content-type":"Text/json"});
                res.end(JSON.stringify(response));
            }
        }
    } else {
        res.write("Page not found");
        res.end();
    }
})

server.listen(port,hostname,()=>{
    console.log(`Server listening at http://${hostname}:${port}`);
});
```
## Screenshot
![image](https://github.com/akash-coded/mern/assets/126975777/f6110cdb-0276-4f1d-b31d-a06bc31eeac3)
![image](https://github.com/akash-coded/mern/assets/126975777/1b0de228-f3e2-4834-b3f6-5f279640262c)
![image](https://github.com/akash-coded/mern/assets/126975777/cc4b9591-1ba6-4d1c-8301-a5c54718fb9d)
![image](https://github.com/akash-coded/mern/assets/126975777/2b45e135-0d2f-4bdc-8719-bbbe39463cde)
![image](https://github.com/akash-coded/mern/assets/126975777/62539533-c3c5-4b96-9ede-8da97da7b2a8)
![image](https://github.com/akash-coded/mern/assets/126975777/71673792-53f5-4509-9a99-cc5f3b9c9e18)

**Comment by vk0027** [comment]
// npm init -y
// npm install http
// Create a new file named index.js in the project directory.
// npm i --g nodemon
// nodemon index.js

const http = require('http');
const url = require('url');

const server = http.createServer((req, res) => {
  const { pathname } = url.parse(req.url);

  if (req.method === 'GET' && pathname === '/') {
    // Display a welcome message and provide instructions for accessing different endpoints
    res.writeHead(200, { 'Content-Type': 'text/plain' });
    res.end('Welcome to the Node.js Server!\n\nAvailable endpoints:\n\n- GET /generate-response\n- POST /process-data');
  } 
  
  else if (req.method === 'GET' && pathname === '/generate-response') {
    // Generate a complex response containing a combination of text and JSON data
    const response = {
      message: 'Welcome to the Complex Response Server!',
      timestamp: new Date().toISOString(),
      data: [
        { name: 'John Doe', age: 30, location: 'New York' },
        { name: 'Jane Smith', age: 28, location: 'San Francisco' }
      ]
    };

    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(response));
  } else if (req.method === 'POST' && pathname === '/process-data') {
    req.on('end', () => {
        const data = JSON.parse(body);
        // Process the received data and generate a response
        const processedData = {
          originalData: data,
          processed: true
        };
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify(processedData));
    } 
)}; else {
    // Handle 404 Not Found errors for undefined routes
    res.writeHead(404, { 'Content-Type': 'text/plain' });
    res.end('404 Not Found');
  }
});

server.listen(5555, () => {
  console.log('Server started on port 5555');
});

**Comment by raihanul-islam-rakin** [comment]
// Import required modules
const http = require('http');

// Create a server
const server = http.createServer((req, res) => {
  // Set response headers
  res.setHeader('Content-Type', 'application/json');

  // Routing
  if (req.url === '/user') {
    const user = { id: 1, name: 'John Doe', email: 'john@example.com' };
    res.end(JSON.stringify(user));
  } else if (req.url === '/products') {
    const products = [
      { id: 1, name: 'Product A', price: 10 },
      { id: 2, name: 'Product B', price: 20 },
    ];
    res.end(JSON.stringify(products));
  } else {
    res.statusCode = 404;
    res.end(JSON.stringify({ message: 'Route not found' }));
  }
});

// Start the server
const PORT = 3000;
server.listen(PORT, () => {
  console.log(`Server is running on port ${PORT}`);
});

**Reply by raihanul-islam-rakin** [reply]
// Start the server
const PORT = 3000;
server.listen(PORT, () => {
console.log(Server is running on port ${PORT});
});
